common/ecdh.c

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2015-2018 Yubico AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+
+#ifdef _WIN32_BCRYPT
+#include <windows.h>
+#include <bcrypt.h>
+#else
+#include <openssl/evp.h>
+#include <openssl/ec.h>
+#include <openssl/ecdh.h>
+#include <openssl/rand.h>
+
+#include "openssl-compat.h"
+#endif
+
+#include "ecdh.h"
+
+#ifdef _WIN32_BCRYPT
+
+// TODO
+
+int ecdh_curve_p256(void) { return 0; }
+int ecdh_calculate_public_key(int curve, const uint8_t *privkey,
+                              size_t cb_privkey, uint8_t *pubkey,
+                              size_t cb_pubkey) {
+  return 0;
+}
+int ecdh_generate_keypair(int curve, uint8_t *privkey, size_t cb_privkey,
+                          uint8_t *pubkey, size_t cb_pubkey) {
+  return 0;
+}
+int ecdh_calculate_secret(int curve, const uint8_t *privkey, size_t cb_privkey,
+                          const uint8_t *pubkey, size_t cb_pubkey,
+                          uint8_t *secret, size_t cb_secret) {
+  return 0;
+}
+
+#else
+
+int ecdh_curve_p256(void) { return NID_X9_62_prime256v1; }
+
+int ecdh_calculate_public_key(int curve, const uint8_t *privkey,
+                              size_t cb_privkey, uint8_t *pubkey,
+                              size_t cb_pubkey) {
+  BN_CTX *ctx = BN_CTX_new();
+  BIGNUM *order = BN_new();
+  BIGNUM *pvt = BN_bin2bn(privkey, cb_privkey, NULL);
```

Comment:
Is it safe to pass `privkey` without a NULL check?

Reply:
No, but the expectation is for higher level functions to protect against that. Found a place where that wasn't the case though. See f21c87dcb6d8b64738a4aa9f54aed3435dfd3632